packages/fxa-auth-server/test/client/index.js

```diff
@@ -826,5 +1060,180 @@ module.exports = (config) => {
+    const resStart = await this.api.passwordChangeStart(
+      this.email,
+      this.authPW,
+      headers || {}
+    );
+
+    this.keyFetchToken = resStart.keyFetchToken;
+    this.passwordChangeToken = resStart.passwordChangeToken;
+
+    if (!this.keyFetchToken) {
+      throw new Error('New key fetch token not returned!');
+    }
+
+    if (!this.kB) {
+      await this.getKeysV1();
+    }
+
+    await this.setupCredentialsV2(this.email, newPassword);
+    await this.derivewrapKbVersion2FromKb();
+
+    const resFinish = await this.api.passwordChangeFinishV2(
+      this.passwordChangeToken,
+      this.authPW,
+      this.authPWVersion2,
+      this.wrapKb,
+      this.wrapKbVersion2,
+      this.clientSalt,
+      headers || {},
+      undefined
+    );
+
+    return resFinish;
+  };
+
+  Client.prototype.getClientSalt = async function(email) {
+    const response = await this.getCredentialsStatus(email || this.email);
+    return response?.clientSalt;
+  }
+
+  Client.prototype.getCredentialsStatus = async function(email, originalLoginEmail) {
+    try {
+      const response = await this.api.getCredentialsStatus(originalLoginEmail || email);
+      return response
+    }
+    catch (error) {
+      return { }
+    }
+  }
+
+  Client.prototype.generateNewWrapKb = function () {
+
+    if (!this.unwrapBKey) {
+      throw new Error('unwrapBKey never determined. Aborting operation.');
+    }
+
+    // When creating new version 2 credentials, we can
+    // create the wrapKb client side and provided it to
+    // the server. This saves us a round trip during
+    // account creation.
+    //
+    // Alternatively we can let the server-side generate the value, and
+    // we can issue a second upgrade credentials call. See v2 account create
+    // test for examples.
+    //
+    this.wrapKb = crypto.randomBytes(32).toString('hex');
+    this.kB = butil.xorBuffers(this.wrapKb, this.unwrapBKey).toString('hex');
+  }
+
+  Client.prototype.deriveWrapKbFromKb = function () {
+
+    if (!this.kB) {
+      throw new Error('kB never determined! Aborting operation.')
+    }
+
+    if (!this.unwrapBKey) {
+      throw new Error('unwrapBKey never determined. Aborting operation.');
+    }
+
+    // By deriving the value this way, we ensure a single kB value. Note that this relies on the
+    // fact the server will be using the current wrapKb for the V1 password
+    this.wrapKb = butil.xorBuffers(
+      this.kB,
+      this.unwrapBKey
+    ).toString('hex');
+  };
+
+  Client.prototype.derivewrapKbVersion2FromKb = function () {
+
+    if (!this.kB) {
+      throw new Error('kB never set. Aborting operation.');
+    }
+
+    if (!this.unwrapBKey2) {
+      throw new Error('unwrapBKey2 never set. Aborting operation.');
+    }
+
+    // By deriving the value this way, we ensure a single kB value. Note that this relies on the
+    // fact the server will be using the current wrapKb for the V1 password
+    this.wrapKbVersion2 = butil.xorBuffers(
```

Comment:
@chenba, this is where we derive wrapKb2, and ensure we don't experience a shift in kB. kB should be private to the client, so this is really only possible here.